packages/react/src/components/NumberInput/NumberInput.tsx

```diff
@@ -255,6 +316,57 @@ const NumberInput = React.forwardRef<HTMLInputElement, NumberInputProps>(
     });
     const [prevControlledValue, setPrevControlledValue] =
       useState(controlledValue);
+
+    const numberParser = useMemo(
+      () => new NumberParser(locale, formatOptions),
+      [locale, formatOptions]
+    );
+    /**
+     * The currently parsed number value.
+     * Only used when type=text
+     * Updated based on the `value` as the user types.
+     */
+    const [numberValue, setNumberValue, isControlled] = useControllableState({
```

Comment:
Adding new state variables instead of using the existing ones is intentional. I tried to be really clear and separate out the `type='number'` and `type='text'` code paths from one another to avoid regressions.